### PR TITLE
Implement CreatePrimaryEx(), refactor CreatePrimary()

### PR DESCRIPTION
**Pull request by twitchy-jsonp**
This implements primitives necessary for remote key attestation for an AIK.

As discussed with awly@, most parameters returned from `CreatePrimaryEx()` are raw byte slices. This is easier to use for their intended use case (remote key attestation), users who just want a primary key have everything they need from `CreatePrimary()` already.

### CODE REVIEW

File: tpm2/tpm2.go [hunk 1]
```diff
@@ -404,7 +404,7 @@ func decodeCreatePrimary(in []byte) (tpmutil.Handle, crypto.PublicKey, error) {
 }
 
 // CreatePrimary initializes the primary key in a given hierarchy.
-// Second return value is the public part of the generated key.
+// The second return value is the public part of the generated key.
 func CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, parentPassword, ownerPassword string, pub Public) (tpmutil.Handle, crypto.PublicKey, error) {
```
Comment (awly) [inline tpm2/tpm2.go]: Refactor this to be a thin wrapper around `CreatePrimaryEx`

File: tpm2/tpm2.go [hunk 2]
```diff
@@ -418,6 +418,57 @@ func CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, par
 	return decodeCreatePrimary(resp)
 }
 
+// CreatePrimaryEx initializes the primary key in a given hierarchy.
+// This function differs from CreatePrimary in that all response elements
+// are returned, and they are returned in relatively raw form.
+//
+// The first return value is a transient handle to the loaded key.
```
Comment (awly) [inline tpm2/tpm2.go]: Use named return values and refer to them by the name instead of position.
Values with obvious names can skip a doc sentence.

File: tpm2/tpm2.go [hunk 3]
```diff
@@ -418,6 +418,57 @@ func CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, par
...
+	cmd, err := encodeCreate(owner, sel, parentPassword, ownerPassword, nil /*inSensitive*/, pub)
+	if err != nil {
+		return 0, nil, nil, nil, Ticket{}, nil, err
+	}
+	resp, err := runCommand(rw, TagSessions, cmdCreatePrimary, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return 0, nil, nil, nil, Ticket{}, nil, err
+	}
+
+	var out struct {
```
Comment (awly) [inline tpm2/tpm2.go]: Move all of this decoding into `decodeCreatePrimary` or `decodeCreate`
Also, you can probably use a bunch of vars instead of a struct

File: tpm2/tpm2.go [hunk 4]
```diff
@@ -418,6 +418,57 @@ func CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, par
...
+		Handle         tpmutil.Handle
+		Public         []byte
+		CreationData   []byte
+		CreationHash   []byte
+		CreationTicket Ticket
+		CreationName   []byte
+	}
+	var paramSize uint32
+
+	offset, err := tpmutil.Unpack(resp, &out.Handle, &paramSize, &out.Public, &out.CreationData, &out.CreationHash)
```
Comment (awly) [inline tpm2/tpm2.go]: use `UnpackBuf` instead of manually handling offsets

File: tpm2/structures.go [hunk 5]
```diff
@@ -792,3 +792,28 @@ type AlgorithmDescription struct {
...
+	Hierarchy uint32
+	Digest    []byte
+}
+
+// Encode represents the given values as a Ticket.
+func (t Ticket) Encode() ([]byte, error) {
+	return tpmutil.Pack(t.Type, t.Hierarchy, t.Digest)
+}
+
+func (t Ticket) packedSize() int {
```
Comment (awly) [inline tpm2/structures.go]: Remove this after you switch to `tpmutil.UnpackBuf` in CreatePrimaryEx

File: tpm2/tpm2.go [hunk 6]
```diff
@@ -461,40 +512,53 @@ func ReadPublic(rw io.ReadWriter, handle tpmutil.Handle) (Public, []byte, []byte
...
-		return nil, nil, fmt.Errorf("decoding CreationData: %v", err)
+		return nil, fmt.Errorf("decoding CreationData: %v", err)
 	}
-	return resp.Private, resp.Public, nil
+	return &resp, nil
 }
 
-// Returns private key and public key blobs.
-func create(rw io.ReadWriter, parentHandle tpmutil.Handle, parentPassword, objectPassword string, sensitiveData []byte, pub Public, pcrSelection PCRSelection) ([]byte, []byte, error) {
+func doCreate(rw io.ReadWriter, parentHandle tpmutil.Handle, parentPassword, objectPassword string, sensitiveData []byte, pub Public, pcrSelection PCRSelection) (*createResponse, error) {
```
Comment (awly) [inline tpm2/tpm2.go]: why rename this?

File: tpm2/tpm2.go [hunk 7]
```diff
@@ -1132,6 +1200,46 @@ func Certify(rw io.ReadWriter, parentAuth, ownerAuth string, object, signer tpmu
...
+	if err != nil {
+		return nil, err
+	}
+	return concat(handles, auth, params)
+}
+
+// CertifyCreation generates a signature of a newly-created &
+// loaded TPM object with a signing key signer. Returned values
+// are: attestation data (TPMS_ATTEST), signature, and an error.
+func CertifyCreation(rw io.ReadWriter, objectAuth string, object, signer tpmutil.Handle, qualifyingData, creationHash []byte, creationTicket Ticket) ([]byte, []byte, error) {
```
Comment (awly) [inline tpm2/tpm2.go]: Can you split this out into a separate PR?

File: tpm2/tpm2_test.go [hunk 8]
```diff
@@ -214,6 +273,36 @@ func TestCombinedEndorsementTest(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("CreatePrimary failed: %v", err)
+	}
+	defer FlushContext(rw, keyHandle)
+
+	pub, _, _, err := ReadPublic(rw, keyHandle)
+	if err != nil {
+		t.Fatalf("ReadPublic failed: %s", err)
+	}
+	pub2, _ := pub.Encode()
```
Comment (awly) [inline tpm2/tpm2_test.go]: handle the error

File: tpm2/structures.go [hunk 9]
```diff
@@ -792,3 +792,28 @@ type AlgorithmDescription struct {
...
+
+// Ticket represents evidence the TPM previously processed
+// information.
+type Ticket struct {
+	Type      tpmutil.Tag
+	Hierarchy uint32
+	Digest    []byte
+}
+
+// Encode represents the given values as a Ticket.
```
Comment (awly) [inline tpm2/structures.go]: `Encode encodes the Ticket into TPM wire format.`

**Comment by twitchy-jsonp** [comment]
PTAL :unicorn: :)

File: tpm2/tpm2.go [hunk 10]
```diff
@@ -353,25 +353,47 @@ func encodeCreateRawTemplate(owner tpmutil.Handle, sel PCRSelection, parentPassw
...
+	}
+
+	if err := tpmutil.UnpackBuf(buf, &creationName); err != nil {
+		return 0, nil, nil, nil, nil, nil, fmt.Errorf("decoding creationName: %v", err)
+	}
+
+	if _, err := DecodeCreationData(creationData); err != nil {
+		return 0, nil, nil, nil, nil, nil, fmt.Errorf("parsing CreationData: %v", err)
+	}
+	return
```
Comment (awly) [inline tpm2/tpm2.go]: Explicitly spell out all return vars, don't do naked returns. It can lead to confusing behavior.

File: tpm2/tpm2.go [hunk 11]
```diff
@@ -461,40 +477,53 @@ func ReadPublic(rw io.ReadWriter, handle tpmutil.Handle) (Public, []byte, []byte
...
 }
 
-func decodeCreate(in []byte) ([]byte, []byte, error) {
-	var resp struct {
-		Handle       tpmutil.Handle
-		Private      []byte
-		Public       []byte
-		CreationData []byte
-	}
+type createResponse struct {
```
Comment (awly) [inline tpm2/tpm2.go]: Remove this and use a long list of returns from `decodeCreate`, like `decodeCreatePrimary`

**Comment by twitchy-jsonp** [comment]
PTAL! :small_airplane:
